File: attractor/agent/skill.py
"""Skill definitions: composable prompt + tool set modifiers for agent nodes."""

from __future__ import annotations

import importlib.util
import logging
from pathlib import Path
from typing import Any

from pydantic import BaseModel, Field

from .tools.base import Tool, ToolRegistry
from .tools import create_default_registry

logger = logging.getLogger(__name__)


def _split_frontmatter(text: str) -> tuple[str | None, str]:
    """Split SKILL.md-style frontmatter from body.

    Returns (frontmatter_yaml, body). Frontmatter is None if the file does not
    open with a `---` delimiter line, or if no closing `---` is found.
    """
    if not (text.startswith("---\n") or text.startswith("---\r\n")):
        return None, text

    lines = text.splitlines(keepends=True)
    for i in range(1, len(lines)):
        if lines[i].strip() == "---":
            return "".join(lines[1:i]), "".join(lines[i + 1 :])
    return None, text


class Skill(BaseModel):
    """A reusable bundle of system prompt additions and tool set modifications."""

    name: str
    description: str = ""
    system_prompt: str = ""
    tools_include: list[str] = Field(default_factory=list)
    tools_exclude: list[str] = Field(default_factory=list)

    model_config = {"arbitrary_types_allowed": True}


class ComposedSkill(BaseModel):
    """Flattened result of merging multiple skills."""

    system_prompt: str = ""
    tools_exclude: list[str] = Field(default_factory=list)
    extra_tools: list[Any] = Field(default_factory=list)

    model_config = {"arbitrary_types_allowed": True}


class SkillRegistry:
    """Loads and composes skills from files and programmatic registration."""

    def __init__(self) -> None:
        self._skills: dict[str, Skill] = {}
        self._custom_tools: dict[str, list[Tool]] = {}  # skill name -> custom tools

    def register(self, skill: Skill, custom_tools: list[Tool] | None = None) -> None:
        """Register a skill programmatically."""
        self._skills[skill.name] = skill
        if custom_tools:
            self._custom_tools[skill.name] = custom_tools

    def get(self, name: str) -> Skill | None:
        return self._skills.get(name)

    def list_skills(self) -> list[Skill]:
        return list(self._skills.values())

    def load_dir(self, path: Path) -> None:
        """Load skills from a directory. Supports .yaml/.yml and .py files."""
        if not path.is_dir():
            logger.warning("Skill directory does not exist: %s", path)
            return

        for entry in sorted(path.iterdir()):
            if entry.suffix in (".yaml", ".yml"):
                self._load_yaml(entry)
            elif entry.suffix == ".py" and not entry.name.startswith("_"):
                self._load_python(entry)

    def _load_yaml(self, path: Path) -> None:
        """Load a skill from a YAML file.

        Supports two formats:
        - Plain YAML mapping (`name: ...`, `system_prompt: ...`, etc.)
        - SKILL.md-style frontmatter: a `---`-delimited YAML block at the top,
          followed by a prose body that becomes `system_prompt`. The skill
          `name` defaults to the file stem if not set in frontmatter.
        """
        try:
            import yaml
        except ImportError:
            logger.warning(
                "PyYAML not installed, skipping skill file: %s. "
                "Install with: pip install pyyaml",
                path,
            )
            return

        try:
            text = path.read_text(encoding="utf-8")
            frontmatter, body = _split_frontmatter(text)

            if frontmatter is not None:
                data = yaml.safe_load(frontmatter) or {}
                if not isinstance(data, dict):
                    logger.warning("Skill file %s frontmatter is not a mapping", path)
                    return
                data.setdefault("name", path.stem)
                if body.strip():
                    data["system_prompt"] = body.strip()
            else:
                data = yaml.safe_load(text)
                if not isinstance(data, dict):
                    logger.warning("Skill file %s does not contain a mapping", path)
                    return

            skill = Skill(**data)
            self._skills[skill.name] = skill
            logger.debug("Loaded skill %r from %s", skill.name, path)
        except Exception as e:
            logger.warning("Failed to load skill from %s: %s", path, e)

    def _load_python(self, path: Path) -> None:
        """Load a skill from a Python module. Expects a module-level `skill` attribute."""
        try:
            spec = importlib.util.spec_from_file_location(f"skill_{path.stem}", path)
            if not spec or not spec.loader:
                logger.warning("Could not load module spec from %s", path)
                return
            module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(module)

            skill = getattr(module, "skill", None)
            if not isinstance(skill, Skill):
                logger.warning("Module %s has no 'skill' attribute of type Skill", path)
                return

            custom_tools = getattr(module, "tools", None)
            if custom_tools and not isinstance(custom_tools, list):
                custom_tools = None

            self.register(skill, custom_tools=custom_tools)
            logger.debug("Loaded skill %r from %s", skill.name, path)
        except Exception as e:
            logger.warning("Failed to load skill from %s: %s", path, e)

    def compose(self, names: list[str]) -> ComposedSkill:
        """Merge multiple skills by name into a single ComposedSkill.

        System prompts are concatenated with newlines. Tool excludes are unioned.
        Custom tools from all referenced skills are collected.
        Unknown skill names are logged as warnings and skipped.
        """
        prompt_parts: list[str] = []
        all_excludes: list[str] = []
        all_extra_tools: list[Tool] = []

        for name in names:
            skill = self._skills.get(name)
            if not skill:
                logger.warning("Unknown skill referenced: %r", name)
                continue

            if skill.system_prompt:
                prompt_parts.append(skill.system_prompt.strip())
            all_excludes.extend(skill.tools_exclude)

            custom = self._custom_tools.get(name, [])
            all_extra_tools.extend(custom)

        return ComposedSkill(
            system_prompt="\n\n".join(prompt_parts),
            tools_exclude=list(dict.fromkeys(all_excludes)),  # dedupe, preserve order
            extra_tools=all_extra_tools,
        )

    def build_tool_registry(self, composed: ComposedSkill) -> ToolRegistry:
        """Build a ToolRegistry from defaults, applying skill modifications."""
        registry = create_default_registry()
        for name in composed.tools_exclude:
            registry.remove(name)
        for tool in composed.extra_tools:
            registry.register(tool)
        return registry

File: internal/agent/skill.go
package agent

import (
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"

	"github.com/nigelpepper/attractor/internal/agent/tools"
)

// Skill is a reusable bundle of system-prompt additions and tool-set changes.
type Skill struct {
	Name         string   `yaml:"name"`
	Description  string   `yaml:"description"`
	SystemPrompt string   `yaml:"system_prompt"`
	ToolsInclude []string `yaml:"tools_include"`
	ToolsExclude []string `yaml:"tools_exclude"`
}

// ComposedSkill is the flattened result of merging multiple skills.
type ComposedSkill struct {
	SystemPrompt string
	ToolsExclude []string
	ExtraTools   []tools.Tool
}

// SkillRegistry loads and composes skills.
type SkillRegistry struct {
	skills      map[string]Skill
	customTools map[string][]tools.Tool
	order       []string
}

// NewSkillRegistry returns an empty registry.
func NewSkillRegistry() *SkillRegistry {
	return &SkillRegistry{skills: map[string]Skill{}, customTools: map[string][]tools.Tool{}}
}

// Register adds a skill programmatically, optionally with custom tools.
func (r *SkillRegistry) Register(s Skill, customTools []tools.Tool) {
	if _, exists := r.skills[s.Name]; !exists {
		r.order = append(r.order, s.Name)
	}
	r.skills[s.Name] = s
	if len(customTools) > 0 {
		r.customTools[s.Name] = customTools
	}
}

// Get returns a skill by name, or (zero, false).
func (r *SkillRegistry) Get(name string) (Skill, bool) { s, ok := r.skills[name]; return s, ok }

// ListSkills returns all skills in registration order.
func (r *SkillRegistry) ListSkills() []Skill {
	out := make([]Skill, 0, len(r.order))
	for _, n := range r.order {
		out = append(out, r.skills[n])
	}
	return out
}

// LoadDir loads skills from a directory (.yaml/.yml files). Python (.py) skill
// modules are not supported in the Go build and are skipped with a warning.
func (r *SkillRegistry) LoadDir(path string) {
	entries, err := os.ReadDir(path)
	if err != nil {
		log.Printf("Skill directory does not exist: %s", path)
		return
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		names = append(names, e.Name())
	}
	sort.Strings(names)
	for _, name := range names {
		full := filepath.Join(path, name)
		switch {
		case strings.HasSuffix(name, ".yaml") || strings.HasSuffix(name, ".yml"):
			r.loadYAML(full)
		case strings.HasSuffix(name, ".py") && !strings.HasPrefix(name, "_"):
			log.Printf("Skipping Python skill module (unsupported in Go build): %s", full)
		}
	}
}

func splitFrontmatter(text string) (string, string, bool) {
	if !strings.HasPrefix(text, "---\n") && !strings.HasPrefix(text, "---\r\n") {
		return "", text, false
	}
	lines := strings.SplitAfter(text, "\n")
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(lines[i]) == "---" {
			return strings.Join(lines[1:i], ""), strings.Join(lines[i+1:], ""), true
		}
	}
	return "", text, false
}

func (r *SkillRegistry) loadYAML(path string) {
	data, err := os.ReadFile(path)
	if err != nil {
		log.Printf("Failed to load skill from %s: %v", path, err)
		return
	}
	text := string(data)
	frontmatter, body, hasFM := splitFrontmatter(text)

	var skill Skill
	if hasFM {
		if err := yaml.Unmarshal([]byte(frontmatter), &skill); err != nil {
			log.Printf("Failed to load skill from %s: %v", path, err)
			return
		}
		if skill.Name == "" {
			skill.Name = strings.TrimSuffix(filepath.Base(path), filepath.Ext(path))
		}
		if strings.TrimSpace(body) != "" {
			skill.SystemPrompt = strings.TrimSpace(body)
		}
	} else {
		if err := yaml.Unmarshal(data, &skill); err != nil {
			log.Printf("Failed to load skill from %s: %v", path, err)
			return
		}
	}
	if skill.Name == "" {
		log.Printf("Skill file %s has no name", path)
		return
	}
	r.Register(skill, nil)
}

// Compose merges named skills: prompts concatenated, excludes unioned, custom
// tools collected. Unknown names are logged and skipped.
func (r *SkillRegistry) Compose(names []string) ComposedSkill {
	var promptParts []string
	var excludes []string
	seenExclude := map[string]bool{}
	var extraTools []tools.Tool

	for _, name := range names {
		skill, ok := r.skills[name]
		if !ok {
			log.Printf("Unknown skill referenced: %q", name)
			continue
		}
		if skill.SystemPrompt != "" {
			promptParts = append(promptParts, strings.TrimSpace(skill.SystemPrompt))
		}
		for _, ex := range skill.ToolsExclude {
			if !seenExclude[ex] {
				seenExclude[ex] = true
				excludes = append(excludes, ex)
			}
		}
		extraTools = append(extraTools, r.customTools[name]...)
	}

	return ComposedSkill{
		SystemPrompt: strings.Join(promptParts, "\n\n"),
		ToolsExclude: excludes,
		ExtraTools:   extraTools,
	}
}

// BuildToolRegistry builds a registry from defaults, applying skill changes.
func (r *SkillRegistry) BuildToolRegistry(composed ComposedSkill) *tools.ToolRegistry {
	registry := tools.CreateDefaultRegistry()
	for _, name := range composed.ToolsExclude {
		registry.Remove(name)
	}
	for _, t := range composed.ExtraTools {
		registry.Register(t)
	}
	return registry
}
